jax/scipy/optimize/bfgs_minimize.py

```diff
@@ -0,0 +1,140 @@
+"""The Broyden-Fletcher-Goldfarb-Shanno minimization algorithm.
+https://pages.mtu.edu/~struther/Courses/OLD/Sp2013/5630/Jorge_Nocedal_Numerical_optimization_267490.pdf
+"""
+
+import jax
+import jax.numpy as jnp
+from jax.lax import while_loop
+from .line_search import line_search
+from typing import NamedTuple, Optional
+
+
+class BFGSResults(NamedTuple):
+  converged: bool  # bool, True if minimization converges
+  failed: bool  # bool, True if line search fails
+  k: int  # The number of iterations of the BFGS update.
+  nfev: int  # The total number of objective evaluations performed.
+  ngev: int  # total number of jacobian evaluations
+  nhev: int  # total number of hessian evaluations
+  x_k: jnp.ndarray  # A tensor containing the last argument value found during the search. If the search converged,
+  # then this value is the argmin of the objective function.
+  f_k: jnp.ndarray  # A tensor containing the value of the objective
+  # function at the `position`. If the search
+  # converged, then this is the (local) minimum of
+  # the objective function.
+  g_k: jnp.ndarray  # A tensor containing the gradient of the objective function at the `final_position`.
+  # If the search converged the l2-norm of this tensor should be below the tolerance.
+  H_k: jnp.ndarray  # A tensor containing the inverse of the estimated Hessian.
+
+
+def bfgs_minimize(func, x0, options=None, _nojit=False):
+  """
+  The BFGS algorithm from
+      Algorithm 6.1 from Wright and Nocedal, 'Numerical Optimization', 1999, pg. 136-143
+
+      Notes:
+          We utilise boolean arithmetic to avoid jax.cond calls which don't work on accelerators.
+          A side effect is that we perform more gradient evaluations than scipy's BFGS
+      func: callable
+          Function of the form f(x) where x is a flat ndarray and returns a real scalar. The function should be
+          composed of operations with vjp defined. If func is jittable then bfgs_minimize is jittable. If func is
+          not jittable, then _nojit should be set to True.
+
+      x0: ndarray
+          initial variable
+      options: Optional dict of parameters
+          maxiter: int
+              Maximum number of evaluations
+          g_tol: flat
+              Terminates minimization when |grad|_2 < g_tol
+          ls_maxiter: int
+              Maximum number of linesearch iterations
+      _nojit: bool
+          Whether to use pythonic control flow so that func without XLA ops can be used. It is also very useful to
+          set _nojit=True in order to perform debugging.
```

Comment:
It is definitely useful to be able to get debugging information, but I think there are cleaner ways to do this than adding an explicit `_nojit=True` option:

- There is a [`disable_jit()` context manager](https://jax.readthedocs.io/en/latest/jax.html?highlight=jax_disable_jit#jax.disable_jit), which basically works exactly like your `_nojit=True`.
- [`jax.experimental.host_callback`](https://jax.readthedocs.io/en/latest/jax.experimental.host_callback.html) allows for getting debug information out, such as intermediate values. This API is still in development but will get better over time.

Reply:
Okay, that's great, I removed these.

Reply:
Done.